design/design-sketch-oam.md

```diff
@@ -0,0 +1,334 @@
+          - name: BACKEND_ADDRESS
+            value: https://example.org
+          livenessProbe:
+            httpGet:
+              port: 8080
+              path: /healthz
+          readinessProbe:
+            httpGet:
+              port: 8080
+              path: /healthz
+    traits:
+    # Contemporary traits model the workload kinds to which they may apply. This
+    # could be modelled as an annotation on the CRD of each Trait kind.
+    - apiVersion: core.oam.dev/v1alpha1
+      kind: ManualScalerTrait
+      spec:
+        replicaCount: 3
+  - component:
+      apiVersion: cache.crossplane.io/v1alpha1
+      kind: RedisCluster
+      spec:
+        classSelector:
+          matchLabels:
+            environment: production
+            region: europe
+        writeConnectionSecretToRef:
+          name: coolest-app-redis
+        engineVersion: "3.2"
+```
+
+Submitting the above `ApplicationController` would cause an instance of each of
+the scopes, components, and traits it templates to be created verbatim (modulo
+some generated object metadata) similar to how the `template` of a Kubernetes
+`ReplicaSet` becomes a set of `Pod` resources. For example:
+
+```yaml
+---
+apiVersion: core.oam.dev/v1alpha1
+kind: ContainerizedWorkload
+metadata:
+  # namespace and name are derived from the parent ApplicationConfiguration.
+  # An owner reference (not modelled here) represents that this scheduling scope
+  # is owned and controlled by its parent ApplicationConfiguration.
+  namespace: default
+  name: coolest-app-f92dm
+  annotations:
+    version: v1.0.0
+spec:
+  # Under this proposal the workloadType would be limited to the core OAM
+  # types, i.e. it would define the function of the resulting workload -
+  # Server, Singleton, Worker, etc. Extended workloads are a distinct kind
+  # and thus kind: ContainerizedWorkload has no workloadSettings field.
+  workloadType: core.oam.dev/v1alpha1.Server
+  osType: linux
+  containers:
+  - name: my-twitter-bot-frontend
+    image:
+      name: example/my-twitter-bot-frontend:v1.0.0
+    resources:
+      cpu:
+        required: 1.0
+      memory:
+        required: 100MB
+    ports:
+    - name: http
+      value: 8080
+    env:
+    - name: USERNAME
+      value: cooladmin
+    - name: PASSWORD
+      valueFrom:
+        secretKeyRef:
+          name: cool-frontend
+          key: password
+    - name: BACKEND_ADDRESS
+      value: https://example.org
+    livenessProbe:
+      httpGet:
+        port: 8080
+        path: /healthz
+    readinessProbe:
+      httpGet:
+        port: 8080
+        path: /healthz
+---
+apiVersion: core.oam.dev/v1alpha1
+kind: ManualScalerTrait
+metadata:
+  namespace: default
+  name: coolest-app-n3v3y
+spec:
+  replicaCount: 3
+---
+apiVersion: cache.crossplane.io/v1alpha1
+kind: RedisCluster
+metadata:
+  namespace: default
+  name: coolest-app-jnkf9
+spec:
+  classSelector:
+    matchLabels:
+      environment: production
+      region: europe
```

Comment:
Ideally this `classSelector` would relate to/be derived from the hypothetical `SchedulingScope`. Doing so would require either:

* The referenced `SchedulingScope` be used to modify the templated `RedisCluster` (which would then differ from its template).
* The `RedisCluster` resource maintain references to any referenced `Scope` resources, and the `RedisCluster` controllers be aware of those scope resources.

Reply:
> Ideally this classSelector would relate to/be derived from the hypothetical SchedulingScope.

Exactly!
In OAM this relationship is declared via AppConfig:

```
kind: AppConfig
metadata: ...
spec:
  components:
  - componentName: ContainerizedWorkload
    instanceName: coolest-app-f92dm
    scopes:
    - name: scheduling-scope-eu-prod
    parameterValues: ...
    traits:
    - type: ManualScalerTrait
      properties:
        replicaCount: 3
```